#!/usr/bin/env bash
#PBS -N combineGvcf
#PBS -o 98_log_files/combine.vcf.err
#PBS -l walltime=50:00:00
#PBS -l mem=15g
#PBS -r n



cd $PBS_O_WORKDIR
Reference="path_to_genome_folder/Pmarg.genome.masked.fa"
index="path_to_genome_folder/Pmarg.genome.masked.dict"
OutDir="04_mapped/gatk"

GATK_jar="gatk-4.0.3.0/gatk-package-4.0.3.0-local.jar"
TmpDir="04_mapped/gatk"
base=__BASE__

#Combine gVCF
java -Xmx10G -Djava.io.tmpdir="$TmpDir" -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar "$GATK_jar" CombineGVCFs \
   -R "$Reference" \
   -O "$OutDir"/combined.raw.gvcf.gz \
	--variant 04_mapped/gatk/sample1.g.vcf.gz \
	--variant 04_mapped/gatk/sample2.g.vcf.gz \
....

## Create vcf output
java -Xmx10G -Djava.io.tmpdir="$TmpDir" -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar "$GATK_jar" GenotypeGVCFs \
   -R "$Reference" \
   -V "$OutDir"/combined.raw.gvcf.gz \
   -O "$OutDir"/combined.raw.vcf.gz
